filesystem/directory_isfs/source/main.c

```diff
@@ -0,0 +1,442 @@
+        __HOST__
+    );
+
+    // Critial Loop.
+    while(true) {
+        // Scan WPAD
+        WPAD_ScanPads();
+        s32 pressed[4]; // State on every remote.
+
+        // Get the remote states.
+        for(int rmt = 0; rmt < 4; rmt++)
+            pressed[rmt] = WPAD_ButtonsDown(rmt);
+
+        // Check for home on each remote.
+        for(int rmt = 0; rmt < 4; rmt++)
+            if(pressed[rmt] & WPAD_BUTTON_HOME || SYS_ResetButtonDown()) break;
+        
+        // Wait VSync.
+        VIDEO_WaitVSync();
+    }
+    exit(code);
+    printf("\nEXITING...\n");
+}
+
+// Determine weather or not an entry is a directory.
+static inline bool is_dir(DIR_ENTRY *entry) {
+	return entry->flags & FLAG_DIR; // If the entry flags has the DIR flag, assume dir.
+}
+
+// Chop off any double slashes (//) from a string.
+static inline void RemoveDoubleSlash(char *str)
+{
+	if(!str) return;
+	int count = 0;
+	const char *ptr = str;
+	while(*ptr != 0)
+	{
+		if(*ptr == '/' && ptr[1] == '/') {
+			ptr++;
+			continue;
+		}
+		str[count] = *ptr;
+		ptr++;
+		count++;
+	}
+
+	while(count > 2 && str[count-1] == '/')
+		count--;
+
+	str[count] = 0;
+}
+
+// Add a child entry to the main parent entry.
+static DIR_ENTRY *add_child_entry(DIR_ENTRY *dir, const char *name) {
+    // If the children section is NULL, allocate one child entry.
+	if(!dir->children)
+		dir->children = malloc(sizeof(DIR_ENTRY));
+    
+    // Reallocate the children array to fit one more child entry.
+	DIR_ENTRY *newChildren = realloc(dir->children, (dir->childCount + 1) * sizeof(DIR_ENTRY));
+	if (!newChildren) return NULL;
+
+    // Zero out memory for additional children entries and update the directory's children pointer.
+	bzero(newChildren + dir->childCount, sizeof(DIR_ENTRY));
+	dir->children = newChildren;
+
+    // Get our new child as a pointer, while incrementing the parent child count.
+	DIR_ENTRY *child = &dir->children[dir->childCount++];
+
+    // Dump the new child name into the actual child.
+	child->name = strdup(name);
+	if (!child->name) return NULL;
+
+    // Allocate space for the child's absolute path.
+	child->abspath = malloc(strlen(dir->abspath) + strlen(name) + 2);
+	if (!child->abspath) return NULL;
+
+    // Set the child's absolute path, and remove double slashes.
+	sprintf(child->abspath, "%s/%s", dir->abspath, name);
+	RemoveDoubleSlash(child->abspath);
+	return child;
+}
+
+// Read a parent directory.
+static bool read_directory(DIR_ENTRY *parent) {
+    // Don't proceed if parent or parent's absolute path is NULL.
+	if(!parent || !parent->abspath)
+		return false;
+    
+    // Have we already read this dir?
+	u32 fileCount;
+	if(parent->size != 0 && is_dir(parent))
+	{
+		fileCount = parent->size;
+    } else
+	{
+        // Read the directory file count.
+		s32 ret = ISFS_ReadDir(parent->abspath, NULL, &fileCount);
+		if (ret != ISFS_OK) {
+			return false;
+		}
+	}
+
+    // Set newly discovered parent parameters.
+	parent->flags = FLAG_DIR;
```

Comment:
i stopped looking at tabbing, maybe just recheck the whole file haha

Reply:
Oh Lord lol, that's weird, the system I submitted the code from doesn't have the tabbing issues, I'll have to investigate that I'm a bit.